libraries/aws_internet_gateway.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+require 'aws_backend'
+
+class AwsInternetGateway < AwsResourceBase
+  name 'aws_internet_gateway'
+  desc 'Verifies settings for an AWS Internet Gateway'
+
+  example "
+    describe aws_internet_gateway(id: 'igw-abc12450edc87a8bd') do
+      it { should be_attached }
+    end
+
+    describe aws_internet_gateway(name: 'my-gateway') do
+      it { should exist }
+    end
+  "
+
+  def initialize(opts = {})
+    super(opts)
+    validate_parameters(require_any_of: %i(id name))
+
+    # Either id or name should be provided.
+    if opts.key?(:id) && !opts.key?(:name)
+      id = opts[:id]
+      format_error = "#{@__resource_name__}: `id` must be a string in the format of "\
+      "'igw-' followed by 8 or 17 hexadecimal characters."
+      pattern = /^igw-[0-9a-f]{8,17}$/
+      raise ArgumentError, format_error unless id.match?(pattern) || id.length.between?(12, 21)
+      query_params = { internet_gateway_ids: [id] }
+    elsif opts.key?(:name) && !opts.key?(:id)
+      name = opts[:name]
+      tag_error = "#{@__resource_name__}: `name` should be maximum 255 characters long and contain "\
+        'letters, numbers, spaces and the following characters only: + - = . _ : / @'
+      raise ArgumentError, tag_error unless name.match?(%r{^[\w+\- =._:/@]{1,256}$}) && name.length.between?(1, 255)
+      query_params = { filters: [
+        name: 'tag:Name',
+        values: [name],
+      ] }
+    else
+      raise ArgumentError, "#{@__resource_name__}: either `id` or `name` must be provided."
+    end
+    @display_name = opts.values.join(' ')
+
+    @failed_resource = false
+    catch_aws_errors do
+      resp = @aws.compute_client.describe_internet_gateways(query_params)
+      gateways = resp.internet_gateways
+      # API will return an empty list if there is no resource with the provided name tag.
+      if gateways.empty?
+        Inspec::Log.warn "#{@__resource_name__}: #{@display_name} not found. "
+        @failed_resource = true
+        return
+      end
+      # Fail resource, redirect users to plural resource if there are multiple resources with the same name tag.
+      if gateways.length > 1
+        fail_resource("#{@__resource_name__}: Multiple internet gateways were returned for the provided parameter:"\
+                      " `#{@display_name}`. If you wish to test multiple internet gateways, "\
+                      'please use the aws_internet_gateways resource.')
+        @failed_resource = true
```

Comment:
Is this instance variable required?

Reply:
It is being used in the `exists?` method.

Reply:
Any reason to use this in the `exists?` method rather than `!gateways.empty?`?

Reply:
`@failed_resource` is `true` if the `gateways.length>1` as well.
It is already declared in the superclass when `catch_aws_erorrs` is used.

There is a `failed_resource?` method defined in the `AwsResourceBase` returning the value of `@failed_resource`, so using that variable in the resource will make this method function properly for the resource.

If I don't use the `@failed_resource`, then I have to make `gateways` an instance variable and move the logic of failing resource into the `exists?` method for the various conditions of `@gateways` variable.
At that point, `@gateway` might be bigger than `true/false`.
I thought it is cleaner to decide if a resource is failed for different scenarios in the `initialize` method and use the outcome in the `exists?` method.

Reply:
The main reason I ask is that there is a similar use of `fail_reosurce` in ec2 instance and I was thinking it would be nice to keep the code as consistent as possible. https://github.com/inspec/inspec-aws/blob/master/libraries/aws_ec2_instance.rb#L46